### ISSUE TITLE
Add support for Cubist models

### ISSUE THREAD

**Issue by edgararuiz** [tidymodels/tidypredict#41]

**Comment by topepo** [comment]
I started this in [my fork](https://github.com/topepo/tidypredict/commit/266df2fd1f8482fb0426ad82d6e5b67ac45209dc). 

The model file is parsed and a set of R expressions are generated. 

The base learner is a model rule-set that has a logical rule and an associated regression equation. For example:

    if
	disp > 95.1
    then
	outcome = 14.89 - 0.0406 hp + 2.4 drat

A rule-set has 1+ rules. 

If committees are used (kind of like boosting), there are multiple rule-sets within a committee. In the example below, there are two committees. The first one has a single rule (that hits all conditions) and the second has two rules. 

I've structured the main part of the output with an R expression so that the `pred_eq` is the product of a logical condition (resulting in zero or one) times the regression equation. 

To make predictions, each new sample would get run through each expression in `pred_eq` (3 in the example below). The prediction is the average of the _active_ rules (which have non-zero values). 

I've put a to-do list in here but maybe we could meet to discuss the best way to go about this. I don't know if my R expressions will translate to SQL easily (e.g. I use `%in%`)

Example:
``` r
library(Cubist)
#> Loading required package: lattice

cars_1 <- cubist(x = mtcars[, -1], y = mtcars$mpg, committees = 2)
summary(cars_1)
#> 
#> Call:
#> cubist.default(x = mtcars[, -1], y = mtcars$mpg, committees = 2)
#> 
#> 
#> Cubist [Release 2.07 GPL Edition]  Sat Mar 30 19:09:39 2019
#> ---------------------------------
#> 
#>     Target attribute `outcome'
#> 
#> Read 32 cases (11 attributes) from undefined.data
#> 
#> Model 1:
#> 
#>   Rule 1/1: [32 cases, mean 20.09, range 10.4 to 33.9, est err 2.29]
#> 
#>  outcome = 37.2 - 3.88 wt - 0.0318 hp
#> 
#> Model 2:
#> 
#>   Rule 2/1: [27 cases, mean 18.09, range 10.4 to 26, est err 2.40]
#> 
#>     if
#>  disp > 95.1
#>     then
#>  outcome = 14.89 - 0.0406 hp + 2.4 drat
#> 
#>   Rule 2/2: [5 cases, mean 30.88, range 27.3 to 33.9, est err 3.77]
#> 
#>     if
#>  disp <= 95.1
#>     then
#>  outcome = 33.06
#> 
#> 
#> Evaluation on training data (32 cases):
#> 
#>     Average  |error|               1.88
#>     Relative |error|               0.40
#>     Correlation coefficient        0.92
#> 
#> 
#>  Attribute usage:
#>    Conds  Model
#> 
#>     50%           disp
#>            92%    hp
#>            50%    wt
#>            42%    drat
#> 
#> 
#> Time: 0.0 secs

source('~/github/forks/tidypredict/R/model-cubist.R')
rule_1 <- parse_model_file(cars_1$model)

# Get per-rule-per-committee prediction equations that are 
# (binary rule condition) * regression equation

rule_1$pred_eq[1:2]
#> [[1]]
#> (TRUE) * (37.2 + (-0.0318 * hp) + (-3.88 * wt))
#> 
#> [[2]]
#> ((disp > 95.099998)) * (14.89 + (-0.0406 * hp) + (2.4 * drat))

# For each sample being predicted, get the contributions (some are zero)
term_contrib <- purrr::map_dbl(
  rule_1$pred_eq,
  rlang::eval_tidy, 
  data = mtcars[1, -1]
)
# All rule results:
term_contrib
#> [1] 23.5364 19.7840  0.0000
# Only activated rules
term_contrib <- term_contrib[term_contrib != 0]
mean(term_contrib)
#> [1] 21.6602

predict(cars_1, mtcars[1, -1])
#> [1] 21.6602
```

Created on 2019-03-30 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)

**Comment by topepo** [comment]
Okay, now that I test a little, I forgot that the prediction equation is a little more complex. There is some [capping of the ranges](https://github.com/topepo/Cubist/blob/f10da8bb6ba4a28aa4c1b4eb85fa7bb6841496da/src/predict.c#L75) as well as [within-committee averaging](https://github.com/topepo/Cubist/blob/f10da8bb6ba4a28aa4c1b4eb85fa7bb6841496da/src/predict.c#L51) but at least it is only a few mutates away:

``` r

library(Cubist)
#> Loading required package: lattice
library(tidyverse)
#> Registered S3 methods overwritten by 'ggplot2':
#>   method         from 
#>   [.quosures     rlang
#>   c.quosures     rlang
#>   print.quosures rlang
#> Registered S3 method overwritten by 'rvest':
#>   method            from
#>   read_xml.response xml2
source('~/github/forks/tidypredict/R/model-cubist.R')

data("attrition", package = "rsample")
attrition <- dplyr::as_tibble(attrition)

cubist_model_commitees <-
  Cubist::cubist(
    x = attrition %>% dplyr::select(-MonthlyIncome),
    y = attrition %>% dplyr::select(MonthlyIncome) %>% unlist(),
    committees = 10
  )

committe_info <- parse_model_file(cubist_model_commitees$model) 

committe_info %>% 
  dplyr::mutate(
    pred_val = purrr::map_dbl(pred_eq, rlang::eval_tidy, data = attrition[1, ]),
    pred_val = ifelse(pred_val == 0, NA, pred_val),
    # Cap the prediction if we are extrapolating
    pred_val = ifelse(pred_val < lower, lower, pred_val),
    pred_val = ifelse(pred_val > upper, upper, pred_val)
  ) %>% 
  group_by(committee) %>% 
  # Committee-level mean
  summarize(comm_mean = mean(pred_val, na.rm = TRUE)) %>% 
  ungroup() %>% 
  # Model-level mean
  summarize(mean = mean(comm_mean, na.rm = TRUE)) %>% 
  pull(mean)
#> [1] 5819.456

predict(cubist_model_commitees, attrition[1,] %>% as.data.frame())
#> [1] 5819.456
```

Created on 2019-03-31 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)

**Comment by edgararuiz** [comment]
@topepo - Hi, thank you so much for looking at this.  Quick question, are you using the Gist I shared with you some time back as the base for this?

**Comment by edgararuiz** [comment]
Addressed by 901df8c8c2275e0a825604591298503126db735a
